modules/c++/six.sicd/include/six/sicd/Utilities.h

```diff
@@ -19,29 +19,140 @@
+                                 const std::vector<std::string>& schemaPaths);
+
+    /*
+     * SicdContents is a helper object for readSicdContents
+     *
+     * This class is a simple wrapper around dynamically allocated
+     * SICD information designed to allow access to either or both 
+     * variables returned by readSicdContents.
+     *
+     */
+    struct SicdContents {
+        // allocated with new
+        ComplexData * complexData;
+        // allocated with new[]
+        std::complex<float>* widebandData;
+    };
+
+    /*
+     * Given a SICD path name and a list of schema, this function reads
+     * and parses the SICD in order to provide the wideband data as well
+     * as the ComplexData associated with the image..
+     *
+     * \param sicdPathname SICD NITF pathname
+     * \param schemaPaths One or more files or directories containing SICD 
+     * schemas
+     *
+     * \return ComplexData associated with the SICD NITF,
+     *           allocated with new
+     *         std::complex<float> buffer containing SICD wideband data,
+     *           allocated with new[]
+     *
+     * \throws See six::sicd::Utilities::getComplexData and 
+     *           six::sicd::Utilities::getWidebandData
+     *
+     */
+    static SicdContents readSicd(const std::string& sicdPathname,
+                                 const std::vector<std::string>& schemaPaths);
+
+    /*
+     * Given a reference to a loaded NITFReadControl, this function
+     * parses the SICD's XML and returns a new ComplexData object.
+     *
+     * \param reader A NITFReadControl loaded with the desired SICD
+     *
+     * \return ComplexData new ComplexData associated with the SICD NITF
+     *
+     * \throws except::Exception if the provided reader is not a SICD
+     *
+     */
+    static ComplexData* getComplexData(NITFReadControl& reader);
+
+    /*
+     * Given a loaded NITFReadControl and a ComplexData object, this 
+     * function loads the wideband data associated with the reader 
+     * and ComplexData object.
+     *
+     * Optionally, a buffer can be provided to read the data into.
+     * If no buffer is provided, this function will allocate its 
+     * own memory with new[].
+     *
+     * \param reader A loaded NITFReadControl associated with the SICD
+     * \param complexData complexData associated with the SICD
+     * \param bufferNumBytes Optional, The size of the given buffer in bytes
+     * \param buffer Optional, A pointer to the buffer to load data into
+     *
+     * \return a pointer to the loaded data.
+     *
+     * \throws except::Exception if the pixel type of the SICD is not a
+     *           complex float32 or complex int64, or
+     *         if the user provided buffer is not large enough to store
+     *           the whole image in, or
+     *         if a buffer size was provided but the buffer pointer is null
+     *
+     */
+    static std::complex<float>* getWidebandData(
```

Comment:
So I agree in principle that this is efficient but it's also easy to accidentally leak memory.  I think there should be two overloadings.  1: you pass in a pre-allocated buffer and it better be big enough to fit or 2: you pass in a `std::vector<std::complex<float>>` which gets resized for you.  The latter can turn around and call the former.  If there are any overloadings where you just pass in a filename (so the caller can't trivially know the size needed upfront since they haven't looked at the ComplexData yet), let's only have a vector version for now.